File: r/src/hangman.r
source('src/wordlist.r')

## Public Functions
startGame <- function(word.list.path) {
    random.word <- getRandomWord(word.list.path)
    if (playGame(random.word)) {
      cat('Congratulations, you guessed the word ', random.word, ' correctly.\n', sep='"')
    } else {
      cat('You did not guess ', random.word, ' correctly, and have been hanged.\n', sep='"')
    }
}

## Privat Functions
playGame <- function(random.word) {

    # Setup
    tries.left <- 10
    guessed.chars <- ''
    current.guess <- paste(replicate(nchar(random.word), '_'), collapse=' ')
    printStatus(current.guess, guessed.chars, tries.left)

    # Game Loop
    while(tries.left != 0) {
        user.guess <- getGuess()
        updated.guessed.chars <- updateGuessedChars(user.guess, guessed.chars)
        updated.current.guess <- updateCurrentGuess(user.guess, current.guess, random.word)

        if ( ! grepl('_', updated.current.guess)) {
            return(TRUE)
        } else if(current.guess == updated.current.guess &&
                  guessed.chars != updated.guessed.chars) {
            tries.left <- tries.left - 1
        }
        current.guess <- updated.current.guess
        guessed.chars <- updated.guessed.chars
        printStatus(current.guess, guessed.chars, tries.left)
    }
    return(FALSE)
}

getGuess <- function() {
  cat('Enter your next guess: ')

  # HACK: allows reading stdin input in batch mode
  con <- if (interactive()) stdin() else file('stdin')
  guess <- scan(file=con, what='character', nmax=1, quiet=TRUE, flush=TRUE)
  close(con)

  return(substr(guess, 1, 1))
}

updateCurrentGuess <- function(user.guess, current.guess, random.word) {
    updated.current.guess <- unlist(strsplit(current.guess, split=' '))

    for (index in 1:nchar(random.word)) {
        char <- substr(random.word, index, index)
        if (char == user.guess) {
            updated.current.guess[index] <- char
        }
    }
    return(paste(updated.current.guess, collapse=' '))
}

updateGuessedChars <- function(user.guess, guessed.chars) {

    # User already guessed that
    if (grepl(user.guess, guessed.chars)) {
        return(guessed.chars)
    }
    return(paste(guessed.chars, user.guess))
}

printStatus <- function(current.guess, guessed.chars, tries.left) {
    cat('Word: ', current.guess, '\n')
    cat('Tries Left:', tries.left, '\n')
    cat('Guessed Letters:', guessed.chars, '\n\n')
    cat('The Gallow\n\n\n')

    if (tries.left == 10) {
        cat('\n\n\n\n\n\n\n')
    } else if (tries.left == 9) {
        cat('\n\n\n\n\n   \n_______________\n\n')
    } else if (tries.left == 8) {
        cat('\n\n\n  |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 7) {
        cat('  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 6) {
        cat('  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 5) {
        cat('  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 4) {
        cat('  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 3) {
        cat('  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 2) {
        cat('  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 1) {
        cat('  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n\n')
    } else if (tries.left == 0) {
        cat('  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n\n')
    }
}

File: clojure/src/hangman.clj
(require 'src.wordlist)

(defn- print-status [current-guess guessed-chars tries-left]
  (println "Word:" current-guess)
  (println "Tries Left:" tries-left)
  (println "Guessed Letters:" guessed-chars "\n")
  (println "The Gallow\n\n")
  (case tries-left
   10 (println "\n\n\n\n\n\n")
    9 (println "\n\n\n\n\n   \n_______________\n")
    8 (println "\n\n\n  |\n  |\n  |\n________________\n")
    7 (println "  |\n  |\n  |\n  |\n  |\n  |\n________________\n")
    6 (println "  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n")
    5 (println "  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n")
    4 (println "  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n")
    3 (println "  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n")
    2 (println "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n")
    1 (println "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n")
    0 (println "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n")))

(defn- guess []
  (print "Enter your next guess: ")
  (flush) ; Needed to ensure a print without a new line is shown
  (let [user-guess-string (read-line)]
    (if (< 0 (count user-guess-string))
      (let [user-guess (first (clojure.string/lower-case user-guess-string))]
        (if (Character/isLetter user-guess)
          user-guess
          (guess)))
      (guess))))

(defn- update-guessed-chars [user-guess guessed-chars]
  (str guessed-chars user-guess " "))

(defn- update-current-guess [user-guess zipped-word-current-guess]
  (reduce (fn [acc elem]
            (if (= user-guess (first elem))
              (str acc (first elem))
              (str acc (second elem)))) "" zipped-word-current-guess))

(defn- string-contains? [chr string]
  (.contains string (str chr)))

(defn- play-game [random-word current-guess guessed-chars tries-left]
  (case tries-left
    0 (print-status current-guess guessed-chars tries-left) false
    :else (do
            (print-status current-guess guessed-chars tries-left)
            (if (= current-guess random-word)
              true
              (let [user-guess (guess)]
                (println) ; Just formatting the output a bit
                (cond
                  (string-contains? user-guess guessed-chars)
                    (play-game random-word current-guess guessed-chars tries-left)
                  (string-contains? user-guess random-word)
                    (play-game random-word
                               (update-current-guess user-guess (map vector random-word current-guess))
                               (update-guessed-chars user-guess guessed-chars) tries-left)
                  :else (play-game random-word current-guess
                                   (update-guessed-chars user-guess guessed-chars) (- tries-left 1))))))))

(defn start-game [word-list-path]
  (let [random-word (read-random-word word-list-path)]
    (if (play-game (clojure.string/lower-case random-word)
                   (apply str (repeat (count random-word) "_")) "" 10)
      (println (str "Congratulations, you guessed the word \"" random-word "\" correctly."))
      (println (str "You did not guess \"" random-word "\" correctly, and have been hanged.")))))
